server/src/main/java/org/elasticsearch/http/AbstractHttpServerTransport.java

```diff
@@ -367,8 +376,13 @@ protected static void onServerException(HttpServerChannel channel, Exception e)
     }
 
     protected void serverAcceptedChannel(HttpChannel httpChannel) {
-        boolean addedOnThisCall = httpChannels.add(httpChannel);
-        assert addedOnThisCall : "Channel should only be added to http channel set once";
+        if (shuttingDown) {
+            logger.warn("server accepted channel after shutting down");
+            httpChannel.close();
+            return;
+        }
+        RequestTrackingHttpChannel trackingChannel = httpChannels.putIfAbsent(httpChannel, new RequestTrackingHttpChannel(httpChannel));
```

Comment:
I think there is a race condition here: out of order instructions could result in the following scenario:

* this method is entered, shuttingDown is checked, reaching the httpChannels put
* shuttingDown set to true
* httpChannels is looped over to set closewhenidle
* this method continues, adding a new channel that does not have closewhenidle set

I think this can be remedied with a second check on shuttingDown after inserting? Another option might be a more complicated httpChannels.computeIfAbsent, where shuttingDown is checked inside the mapping lambda. Some flag would then need to be set to indicate the need to return immediately. There may be other ways to fix this race condition as well.

Reply:
I have added a read write lock around `shuttingDown` which should handle the race.